### DISCUSSION TITLE
Strange ugly black edges of text after using paste

### DISCUSSION THREAD

**Discussion by KimigaiiWuyi** [python-pillow/Pillow#8359] [Q&A] [answered]
```python
from PIL import Image, ImageDraw, ImageFont

font = ImageFont.truetype(r'C:\Users\44483\Desktop\zzz_thins.ttf', size=60)
img = Image.new("RGBA", (200, 200), (200, 160, 130, 0))
img_draw = ImageDraw.Draw(img)
img_draw.text((0, 0), "你好", font=font, fill=(255, 255, 255))

base = Image.new("RGBA", (200, 100), (0, 0, 0, 120))
base_draw = ImageDraw.Draw(base)
base_draw.text((0, 0), "你好", font=font, fill=(255, 255, 255))
base.save("test1.png")

img.paste(base, (0, 100), base)
img.save("test2.png")
```
![test1](https://github.com/user-attachments/assets/ab749c0c-39e4-461c-89bb-1469084e31ab)
![test2](https://github.com/user-attachments/assets/6e5f7391-4abb-4f75-b3fe-04770eb0644a)

After paste, the background color seems to fade and there is a strange black edge around the text，
This looks very strange and performs worse on lower pixel plots！

![test3](https://github.com/user-attachments/assets/67cc7741-b5dd-4cef-8212-2aa48edcbb88)

I'm not sure what's going on and I'm hoping someone can answer my questions for me！extremely grateful！

**Comment by radarhere** [comment]
I can't see the 'strange black edge' in the images you've attached.

As to why the background colour is faded, you have provided `base` as the third argument to [`paste()`](https://pillow.readthedocs.io/en/stable/reference/Image.html#PIL.Image.Image.paste), making it the 'mask'.

https://pillow.readthedocs.io/en/stable/reference/Image.html#PIL.Image.Image.paste
> If a mask is given, this method updates only the regions indicated by the mask.

Because `base` image has a background of 120 alpha, the pasted image is faded. To demonstrate, if you reduce
```python
base = Image.new("RGBA", (200, 100), (0, 0, 0, 120))
```
to
```python
base = Image.new("RGBA", (200, 100), (0, 0, 0, 0))
```
you will see the background disappear entirely.

If you do not want the background to fade, then you can just change
```python
img.paste(base, (0, 100), base)
```
to
```python
img.paste(base, (0, 100))
```
to stop using a mask in your call to `paste()`.

**Reply by KimigaiiWuyi** [reply]
Maybe it's just my words, but when I draw text, I always see a strange “overflow” around the edges of the text, which may be due to the fact that PILLOW can't antialiased the text, or it may be due to something else.
Take a look at my screenshot of the finished product, the text in the six groups of areas below, the edges of the text generated by PILLOW are always not very clear, the edges of the text will be fuzzy, I'm not quite sure what this is due to.

It's like a Matte Choker.

https://github.com/user-attachments/assets/e602158f-909e-4542-84d2-9dbbd931d180
comparison

**Reply by radarhere** [reply]
Unfortunately, the further images you've provided don't clarify anything for me, as they introduce many more pieces of text into the conversation.

Could you take the image from your first comment, and manually 'correct' it with Photoshop or similar to demonstrate what you want?

**Reply by KimigaiiWuyi** [reply]
> Unfortunately, the further images you've provided don't clarify anything for me, as they introduce many more pieces of text into the conversation.
> 
> Could you take the image from your first comment, and manually 'correct' it with Photoshop or similar to demonstrate what you want?

Of course, note the difference in the edges between the two.
The text generated by pillow has a very thin (about 1px) shadow stroke, which should not normally look like this.

![Snipaste_2024-09-08_14-50-19](https://github.com/user-attachments/assets/78c98d3d-d409-45e2-a7e9-5b769edddf7c)

**Reply by radarhere** [reply]
Your original code doesn't paste the image onto a purple background though. If you're taking further steps after your original code that you're not showing here, it's hard to help fix that.

Have you tried the code I suggested earlier? Maybe that will solve your second problem as well?

Could you point us to a copy of zzz_thins.ttf?

**Reply by KimigaiiWuyi** [reply]
> 但是，您的原始代码不会将图像粘贴到紫色背景上。如果您在原始代码之后采取了未在此处显示的进一步步骤，则很难帮助修复该问题。
> 
> 您尝试过我之前建议的代码吗？也许这也能解决您的第二个问题？
> 
> 您能给我们提供一份 zzz_thins.ttf 的副本吗？

Of course, I can provide the font files, the purple background is for comparison, in fact any background color will have a stroke of about 1px, if the font size is bigger, it won't have much effect, but if the font size is smaller than 25, this stroke will affect the body of the font very much.

[zzz_thins.zip](https://github.com/user-attachments/files/16938513/zzz_thins.zip)

**Reply by radarhere** [reply]
If I remove the mask from your code
```python
from PIL import Image, ImageDraw, ImageFont

font = ImageFont.truetype('zzz_thins.ttf', size=60)
img = Image.new("RGBA", (200, 200), (200, 160, 130, 0))
img_draw = ImageDraw.Draw(img)
img_draw.text((0, 0), "你好", font=font, fill=(255, 255, 255))

base = Image.new("RGBA", (200, 100), (0, 0, 0, 120))
base_draw = ImageDraw.Draw(base)
base_draw.text((0, 0), "你好", font=font, fill=(255, 255, 255))
base.save("test1.png")

img.paste(base, (0, 100))
img.save("test2.png")
```
the 'fade' on the background colour is fixed in test2.png.

![test2](https://github.com/user-attachments/assets/42398298-5491-4971-807d-3c393351d59d)

If I modify it to have purple and blue backgrounds, I don't think I see the 'strange ugly black edges' that you describe.
```python
from PIL import Image, ImageDraw, ImageFont

font = ImageFont.truetype('zzz_thins.ttf', size=60)
img = Image.new("RGBA", (200, 200), "#f0f")
img_draw = ImageDraw.Draw(img)
img_draw.text((0, 0), "你好", font=font, fill=(255, 255, 255))

base = Image.new("RGBA", (200, 100), "#0ff")
base_draw = ImageDraw.Draw(base)
base_draw.text((0, 0), "你好", font=font, fill=(255, 255, 255))
base.save("test1.png")

img.paste(base, (0, 100))
img.save("test2.png")
```
<img src="https://github.com/user-attachments/assets/1fe1f336-916d-4c14-9e84-6259368337ba" width="500" />

**Comment by KimigaiiWuyi** [comment]
I'm sorry to say that I've reorganized a test case, in which it is obvious that the text generated by the pillow seems to be blocked by a circle (about 1px or less). @radarhere 

pillow(right)
Photoshop(left)

![sd](https://github.com/user-attachments/assets/dbe6ed9b-9da4-4e45-9d3a-403aeb4f79ea)

```python
from pathlib import Path

from PIL import Image, ImageDraw, ImageFont

path = Path(__file__).parent

font = ImageFont.truetype(str(path / 'zzz_thins.ttf'), size=15)

img1 = Image.open(path / '1.png')
img2 = Image.open(path / '2.png')
img2_draw = ImageDraw.Draw(img2)
img2_draw.text((125, 50), '你好测试', 'white', font, 'mm')

img_temp = Image.new('RGBA', img1.size)
img_temp.paste(img2, (75, 150))

img1 = Image.alpha_composite(img1, img_temp)


img1.save(path / 'test.png')

```
[ex.zip](https://github.com/user-attachments/files/16945052/ex.zip)

![ex](https://github.com/user-attachments/assets/b7b33cac-432d-4469-88c5-1307754428d5)

**Reply by radarhere** [reply]
You're referring to the third character (测) looking better in the Photoshop version?

**Reply by KimigaiiWuyi** [reply]
> You're referring to the third character (测) looking better in the Photoshop version?

almost each character, and the more strokes there are, the more it feels like a portion of the surrounding area has been subtracted, like a stroke or an overshoot.

Simply put, at a glance, the text will look more faded and grayed out in comparison.

I'm not sure you understand what I'm saying.
Maybe it's a simple subtle difference, but in some cases and contexts, the text display and rendering will look worse than in my simple example.

![sd2](https://github.com/user-attachments/assets/c08b459c-1ab5-47f8-ad90-f3f91b30d2f6)

**Reply by radarhere** [reply]
Ok, so to be clear, this problem has nothing to do with pasting anymore.

I might be able to figure out a PR to allow for fractional stroking of the text, e.g. `img2_draw.text(..., stroke_width=0.25)`. Putting together a quick demo, would this output be sufficient to resolve this?

![test](https://github.com/user-attachments/assets/bf7ae53d-dafc-42af-ae63-e4dda4997c68)

**Reply by KimigaiiWuyi** [reply]
Thank you very much!  😘 Bolding all characters synchronously by 0.25px seems to cause some stroke sticking issues, but it also solves part of the problem.
What causes the difference in character display and rendering between the two above? Is there a way to fix it?

**Reply by radarhere** [reply]
Pillow uses [FreeType](https://freetype.org/) as the foundation for text rendering.

Photoshop is commercial software, so I don't know exactly what goes on inside - but I doubt it uses FreeType. I found an early comparison between [FreeType and Photoshop output](https://freetype.org/freetype1/screenshots.html), which would hardly have been made if Photoshop was just using FreeType under the hood.

I don't think 'the output doesn't look like Photoshop' is a concrete problem that can be addressed. If you have a more definitive reason to believe that characters are being rendered incorrectly, that is not just a result of the font, please let us know.

If you don't like how Pillow renders the font at small sizes, my only suggestion is to try rendering the font at a larger size, and then resize the resulting image.

**Reply by KimigaiiWuyi** [reply]
Really appreciate your answer, I fully understand everything, I'm looking forward to being able to use float values for stroke width, this should improve things a lot, thanks for your help!😘

**Reply by radarhere** [reply, marked as answer]
I've created #8369 to accept float stroke widths.
